cpmpy/transformations/linearize.py

```diff
@@ -139,10 +144,29 @@ def linearize_constraint(lst_of_expr, supported={"sum","wsum"}, reified=False):
                             raise NotImplementedError("Modulo with a divisor domain containing 0 is not supported. Please safen the expression first.")
                         k = intvar(*get_bounds((x - rhs) // y))
                         mult_res, newcons = get_or_make_var(k * y)
-                        newlist += linearize_constraint([rhs < abs(y)]+newcons, supported, reified=reified)
+                        # (abs of) modulo rhs is smaller than (abs of) the divisor y, but also needs to be of same sign as x.
+                        newlist += linearize_constraint(only_implies(only_bv_reifies(flatten_constraint([Abs(rhs) < Abs(y), (x > 0).implies(rhs >= 0), (x < 0).implies(rhs <= 0)]))) + newcons, supported, reified=reified)
```

Comment:
Should we make this nicer? also confused why we don't have to decompose the abs?

Reply:
Well, I'm not sure what this is doing for currently?
It also ties back to the division-saga. 

As we decided to do integer division, do we need to decide how to implement Modulo too? Or is this independent?

Reply:
conceptually it's dependent but you can implement it independantly. You can define modulo as the remainder after floordivision (which is how it's mostly defined mathematically I think), but our solvers (using integer division) seem to define modulo as the remainder after integer division (so we mimic this to stay in line with the our integer division decision, also only z3 does it different but also not according to the mathematical definition as far as i can tell)

Reply:
this line says the remainder has to be smaller than the divisor (otherwise we could fit another time the divisor). Had to add the constraints that determine the sign of the remainder, otherwise you can always have multiple solutions (one negative and one positive).

Reply:
but these constraint have to go through our transformations again.. I guess we could manually linearize them as well